sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/FileOpenInputStreamResult.java

```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.file.datalake.models;
+
+import java.io.InputStream;
+
+/**
+ * Result of opening an {@link InputStream} to a datalake file.
+ */
+public interface FileOpenInputStreamResult {
```

Comment:
should this be `Closeable` since it wraps closeable stream?

Reply:
I'm not sure what we gain from this since they have to get the InputStream anyway. I guess someone who only got the properties would benefit, but they might be writing some weird code if that's where they find themselves. Most people working with streams are probably used to the stream being closable and adding a second hook to close the same thing may be confusing.

Reply:
We do that in dotnet, see https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/src/Models/BlobDownloadStreamingResult.cs , i.e. a model holding reference to Disposable type should be Disposable itself.

@alzimmermsft is there a guideline around models that has `Closeable` property inside ?

Reply:
As far as I know there is no Azure SDK for Java guideline about implementing `Closeable`/`AutoClosable` if underlying properties implement those interfaces. That being said, if the expectation for this type is to be used alongside other stream types I'd implement `AutoClosable` to allow for try-with-resources functionality.

cc: @srnagar

Reply:
I'll add closable for now and leave it to final api review to catch this if it's undesirable.

Reply:
Closing a stream more than once is not always idempotent - https://bugs.openjdk.java.net/browse/JDK-8054565

Users are more likely to use try-with-resource when they call the get method - `try(InputStream stream = result.getInputStream()){ } `. 

I don't expect users to use try-with-resources on a type named *Result. So, I am okay with this type not implementing `Closeable`/`AutoCloseable` and document that the stream returned by `getInputStream()` should be closed just because not all streams may have an idempotent `close()` method.

The only implementation I see for `FileOpenInputStreamResult` is the `InternalFileOpenInputStreamResult` which overrides `close()` method and is safe to call twice but we may add more implementations later.

Reply:
@srnagar What if then never call `getInputStream()`? with current implementation the stream sitting inside the wrapper is "opened". I.e. if somebody writes code
```
{
   FileOpenInputStreamResult  result = client.openInputStream();
   System.out.println(result.getProperties().getSomeProperty())'
}
```
That snippet will leave "opened" stream in memory until GC collects it and there's no indication that user should close the stream sitting inside.
The issue about idempotency you mention is classified as bug - so under normal circumstances should we expect close to be idempotent?

Reply:
Making `FileOpenInputStreamResult` implement `Closeable` is still not going to fix the issue if the code is written as above. 

The name `*Result` doesn't clearly indicate that this is a stream that needs closing.

Couple of options here:
- Rename `FileOpenInputStreamResult` to `FileOpenInputStream` and implement `AutoCloseable`
- Rename `FileOpenInputStreamResult` to `FileOpenInputStream` and extend `BlobInputStream` 

Also, just curious why this is type is an interface? Do we expect more implementations to be added later?

Reply:
@srnagar I understand that making it closeable won't make the code snippet work, but it will make people who pay attention or using static analyzers to notice they should add try-with-resources - without closeable they won't get that hint.

I'm not a big fan of both alternatives. This type isn't a stream and we don't want to mix datalake and blob models. Neither we want to create stream derivative just to smuggle few extra properties.

If we're to make a trade-off here then I'd say we remove `Closeable` because the implementation of stream we return anyway has no-op close and it's highly unlikely it's going to change. If that changes we can make getInputStream to lazy-load or create new stream each time it's called.

The reason this type is interface is because we want to hide concrete implementation inside "internal" package. With ever expanding models it's easier to maintain existing types if they're not exposed "publicly". This is lesson learnt from [here](https://github.com/Azure/azure-sdk-for-java/blob/70e7db4426fe6a9b69f20f3a39a7a2dd09982828/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobProperties.java#L64-L434) - we don't want to add new ctor overload everytime we add property to the immutable type.